tests/MapTest.php

```diff
@@ -208,5 +208,21 @@ public function testTextAsKey() {
 		$map->remove($key);
 		$this->assertEquals(0, $map->size());
 	}
+
+    public function testCloneWithObjects() {
+        $obj1 = (object) ['id' => 1, 'name' => 'Joda'];
+        $obj2 = (object) ['id' => 2, 'name' => 'Obi Wan'];
+        $map = new Map(['obj1' => $obj1, 'obj2' => $obj2]);
+        $clone = clone $map;
+
+        $this->assertTrue($clone instanceof Map);
+        $this->assertEquals($map, $clone);
+        $this->assertNotSame($map, $clone);
+
+        $this->assertEquals($map->get('obj1'), $clone->get('obj1'));
+        $this->assertNotSame($map->get('obj1'), $clone->get('obj1'));
+        $this->assertEquals($map->get('obj2'), $clone->get('obj2'));
+        $this->assertNotSame($map->get('obj2'), $clone->get('obj2'));
```

Comment:
This test is missing `assertNotSame()` for primitives. I would assume with the above implementation, that this will be true:

```
$list = new ArrayList('1', 2);
$clone = clone $list;

$this->assertSame($list->get(1), $clone->get(1));
```

however, they shouldn't - should they?

Reply:
Btw: I'm a bit rusted of how php manages references here (that would be true for JS, which is what I'm programming in for the past years).

Reply:
Yes, `assertNotSame` for primitives always fails and it's the expected PHP behavior.
Php automatically creates a copy of the primitives, with different references but afaik there isn't a way to test it.